### ISSUE TITLE
__dunder__ + @functools.singledispatchmethod results in method not defined on type

### ISSUE THREAD

**Issue by ArthurBook** [microsoft/pyright#7358]
**Describe the bug**
decorating a `__dunder__` with `@functools.singledispatchmethod` results in `method not defined on type`

**Code or Screenshots**
```python
import functools
from typing import Union


class Test:
    @functools.singledispatchmethod
    def __getitem__(self, key: Union[int, str]) -> Union[int, str]: ...

    @__getitem__.register
    def get_int(self, key: int) -> str:
        return "a"

    @__getitem__.register
    def get_str(self, key: str) -> int:
        return 0


# works in runtime
bing = Test()
print(bing[1])  # ""
print(bing["1"])  # 0

bing[1]  # "__getitem__" method not defined on type "Test"PylancereportIndexIssue
```

```bash
$ pyright test.py 
test.py
test.py:20:7 - error: "__getitem__" method not defined on type "Test" (reportIndexIssue)
test.py:21:7 - error: "__getitem__" method not defined on type "Test" (reportIndexIssue)
test.py:23:1 - error: "__getitem__" method not defined on type "Test" (reportIndexIssue)
3 errors, 0 warnings, 0 informations 
```

**VS Code extension or command-line**
pyright in CLI:
 version      : 1.1.351                          

dependencies:
 - nodeenv >=1.6.0

**Comment by erictraut** [comment]
Pyright has no specialized support for singledispatch currently, so it treats it like any other method. We currently have no plans to add support for singledispatch, but it's something we might consider in the future if there's sufficient interest (in the form of upvotes). I'm going to close this issue for now, but I'll revisit it in the future if there are enough reactions.
